feat: basic model listing / search page

leonardmq opened the pull request:
## What does this PR do?

Add a new (very much open to changes) page to list and search through available models:
- Uses the remote model list (may want to have it go through the backend API to have cleaner typing?)
- Search by model slug or name (non-fuzzy for now)
- Filter by provider, model family, and other criteria
- Summary view per model showing:
  - Which providers support it
  - What features each provider supports

### Why

Model discovery currently happens on external providers like OpenRouter. Users then need to manually map those models back to Kiln, where the names differ (OpenRouter gives a slug; and it is hard to know for sure which model it is in Kiln in some cases where the model is a narrow variant).

We can build on top of this later to let users click a model to go directly into the `Run` UI with the right model and provider selected or other such things.

https://github.com/user-attachments/assets/7589e14a-0dac-4aef-ad31-e4093babb549




<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Introduced a "Models" section in the app navigation, accessible from the sidebar.
  * Added a comprehensive Model Library page with search, filtering, and sorting capabilities for models.
  * Visual indicators for provider connection status and detailed model information, including capability badges and tooltips.

* **Bug Fixes**
  * Loading and error states are now handled gracefully with informative messages and retry options.

* **Chores**
  * Expanded the list of recognized model names to include new models.
  * Disabled server-side rendering for the Models page to enhance client-side performance.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

scosman commented:
This is great. Some small nits:

Can we make "suggested" models pop a little at the top level?
 - "Data Gen" Badge could be "Data Gen ★" or "Data Gen +" for suggested, with the tool tip being more verbose on what's special "Recommended for Data Gen - one of the best models."

We should add "uncensored" and an icon for "suggested_for_uncensored"

Tooltips need a little work:
 - "Supports Data Gen" -> "Supports synthetic data gen: tested with Kiln's data gen tool."
 - "Supports structured output" -> "Supported structured output (JSON) preferring the tool-calling method".
 - "Support Logprobs" -> "Supports logprobs, a feature needed for the advanced eval method G-Eval".

title "Models" "Model Library"
Subtitle: "Browse available models"

scosman commented:
We should have a UI indicator if I don't have a provider connected. Something subtle. just color ideally, great out logo/title a bit?

Clicking should take you to connect_providers

scosman commented:
"No models found" SVG needs some work. Sonnet can't draw.

Sidebar icon should be: https://www.svgrepo.com/svg/524675/library

scosman commented:
Let's also add an AppPage button: "Manage Providers" to `/settings/providers`? Just grey, not primary.

leonardmq commented:
@scosman 

Right now, we are (probably) dealing with untyped strings on the frontend, but we cannot rely on that - we expose `ModelProviderName` as a type and already use it: 
https://github.com/Kiln-AI/Kiln/blob/96292cb270057b03ec9e1d35529ee210b0c3204b/app/web_ui/src/lib/types.ts#L41

Typescript types are not enforced at runtime so this would not be immediately obvious that things are breaking (unlike with Python enums), but any code assuming `ModelProviderName` is exhaustive could break. For example, provider logos defined here would require a fallback / default state to handle unknown providers, and this would not be caught in typing at compile time: 
https://github.com/Kiln-AI/Kiln/blob/96292cb270057b03ec9e1d35529ee210b0c3204b/app/web_ui/src/lib/ui/provider_image.ts#L3

Probably safer to just serve the config via an endpoint on our API to filter out the ones that backend also does not support. Later we could also add a flag to indicate to the client that the config it is getting excludes some models that are not compatible and that they should try to update Kiln.

leonardmq commented on app/web_ui/src/routes/(app)/models/+page.svelte:
```diff
@@ -0,0 +1,753 @@
+    }
+
+    return false
+  }
```
Thanks - refactored in https://github.com/Kiln-AI/Kiln/pull/439/commits/7530e71d5e39bc04ddcbfd97ca00b8ae38253a95

scosman commented:
# 1
Everything seems to be getting the special star. We're treating "supports_data_gen" as if it's "suggested_for_data_gen". Maybe the comment wasn't clear:

 - it should still have "Data Gen" badge if supports_data_gen
 - it should be "Data Gen [star]" if suggested_for_data_gen

# 2
Same goes for other badges like uncensored. 

# 3 
Log probs is missing a badge

# 4 
Can we make the "Evaluations" item in dropdown "Suggested For Evals". Technically all models support evals, we want to make it clear this is a "suggestion" filter

# 5
I'll maybe to this separately, but we're not listing most of our Fine-tune models. They come direct from the Fireworks API, and don't have nice names so I don't just want to drop them in. I might add a modal or something.

leonardmq commented:
@scosman 

That was partially intentional; I kept the badges only for the starred traits (there was a bug where `supports_data_gen` was one of them, instead of `supports_data_gen`); and the idea was to keep the more granular details at the provider level.

Changes:
- fix: the star badge is `suggested_for_data_gen` (mistakenly set to be `supports_data_gen`)
- added back all the traits (as opposed to only star ones) as badges; including uncensored and logprobs
  - if a model is both `Suggested for XYZ` and `XYZ`, then we only show `Suggested for XYZ`
  - star badges are listed first
- rename capability option into `Suggested For Evals` 
- changed some colors; one was `yellow` and when displayed as a colored circle next to the providers, it looked too much like a status indicator suggesting a warning

Regarding `5.`, confirming we will do that separately.

Also filed this Issue: https://github.com/Kiln-AI/Kiln/issues/475 - will address separately as separate issue

leonardmq commented on app/web_ui/src/routes/(app)/models/+page.svelte:
```diff
@@ -0,0 +1,814 @@
+    supports_finetuning:
+      "Supports finetuning, to train the model on your own use cases.",
+    supports_data_gen: "Supports Synthetic Data Generation.",
+    supports_structured_output:
```
I find this message confusing - not sure what `preferring the tool-calling method` means in this context
